File: py/linear_model.py
import numpy as np


def optimize(x, y):
    x = convert_to_mat_with_x0(x)
    params = np.dot(np.dot(np.dot(np.transpose(x), x).I, np.transpose(x)), y)
    return params


def calc_value(x, params):
    x = convert_to_mat_with_x0(x)
    y_hat = np.dot(x, params)
    return y_hat


def convert_to_mat_with_x0(x):
    return np.c_[np.ones(x.shape[0]), x]

File: src/main/java/jm/app/algorithm/LinearModel.java
package jm.app.algorithm;

import jm.app.optimize.BasicModel;
import jm.app.algebra.AlgebraUtil;
import jm.app.algebra.Matrix;

import java.math.BigDecimal;

public class LinearModel implements BasicModel {

    @Override
    public Matrix optimize(Matrix x, Matrix y) {
        Matrix newX = convertToMatrixWithX0(x);
        Matrix newXT = AlgebraUtil.transpose(newX);
        Matrix params = AlgebraUtil.multiply(AlgebraUtil.multiply(AlgebraUtil.inverse(AlgebraUtil.multiply(newXT, newX)), newXT), y);
        return params;
    }

    @Override
    public Matrix calcValue(Matrix x, Matrix params) {
        Matrix newX = convertToMatrixWithX0(x);
        Matrix yHat = AlgebraUtil.multiply(newX, params);
        return yHat;
    }

    private Matrix convertToMatrixWithX0(Matrix x) {
        Matrix newX = new Matrix(x.getRowNum(), x.getColNum() + 1);
        for (int i = 1; i < newX.getColNum(); i++) {
            for (int j = 0; j < newX.getRowNum(); j++) {
                BigDecimal value = new BigDecimal(x.getValue(j, i - 1).doubleValue());
                newX.setValue(j, i, value);
            }
        }
        for (int i = 0; i < newX.getRowNum(); i++) {
            newX.setValue(i, 0, 1);
        }
        return newX;
    }

    @Override
    public String toString() {
        return "Linear Model";
    }
}
